### ISSUE TITLE
[JENKINS-10912] Browser/JS memory leak in dashboard

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#11185]
<p>leaving the dashboard webpage open (with autorefresh) will accumulate lots of memory.  I left it open over the weekend and it was using over 700 MB of ram!  you can watch the memory consumption increase ~3-10 MB every time it refreshes (manually or automatically).  Sometimes it will decrease to the pre-refreshed amount, but it will always increase over time (instantaneous readings may/may not prove true, but long term averages will).</p>

**Attachments:**

- [OoM.jpg](https://issues.jenkins.io/secure/attachment/21805/OoM.jpg)
- [mem_1.457.jpg](https://issues.jenkins.io/secure/attachment/21826/mem_1.457.jpg)
- [mem_1.460.jpg](https://issues.jenkins.io/secure/attachment/21825/mem_1.460.jpg)

---
<details><summary><i>Originally reported by <a title="david abadir" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abadird">abadird</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-10912" target="_blank">Browser/JS memory leak in dashboard</a></i></summary>
<i><ul>
<li><b>reported by</b>: abadird
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-10-14T22:45:10+00:00
<li><b>votes</b>: 20
<li><b>watchers</b>: 24
<li><b>imported</b>: 2025-11-07
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=elik">elik</a>:</i>
<p>I've noticed memory inflation of 1GB overnight with our Jenkins, and the issue seems to be present on the demo jenkins (<a href="http://ci.jenkins-ci.org/" class="external-link" target="_blank" rel="nofollow noopener">http://ci.jenkins-ci.org/</a>). I didn't leave it open for that long, but it did increase by 10MB over 10-15 minutes.</p>

<p>I use Chrome 17 (beta) and tried to see where all that memory was going with dev tools. I tried taking a heap snapshot of my 1GB tab but it crashed it. I tried again after just 10MB of "leak" but couldn't notice much difference in components size. I did notice that memory fluctuates whenever there is network traffic, which occurs when there are ajax requests (ajaxBuildQueue, ajaxExecutors). Those are triggered whether auto refresh is enabled or not (in my case, auto refresh were disabled).</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=lneff">lneff</a>:</i>
<p>I've noticed this on Jenkins 1.460, running on Windows, I'm using FireFox.  My Jenkins instance seems to be constantly growing in size, I hoped upgrading to 1.460 from approx. 1.420 would help, but doesn't seem to be doing so.  This morning I notice that with 1.460, and autorefresh off, just navigating from one Jenkins page to another causes memory usage to increase several K, and space never seems to be recovered.  Our Jenkins instance is quiet this morning, no jobs running.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=akarollil">akarollil</a>:</i>
<p>We are seeing this problem too (on v1.461). Over time Jenkins uses about 700MB for about 10 jobs. The 'Monitoring' plugin (quite useful by the way) lets you do a garbage collection but that cleared up only 30MB. The histogram shows about 150MB usage over the past week and then a quick ramp yesterday from 150MB to 700MB. There was a jump in number of sessions from 3 to 9 around the time just before the ramping in memory usage happened. I am not sure if its related and the number of sessions dropped to 4 a bit after. CPU usage is also quite high right now, mostly tying down one CPU core.</p>

<p>Memory histogram shows this:</p>

<p>Class Size (Kb) % size Instances % instances<br/>
byte[]	132,623	17	179,202	1	 <br/>
char[]	125,861	16	1,630,859	9	 <br/>
java.util.HashMap$Entry	65,111	8	2,778,092	16	 <br/>
java.util.HashMap$Entry[]	51,225	6	700,897	4	 <br/>
java.lang.String	43,672	5	1,863,370	11	 <br/>
java.lang.Object[]	30,612	4	453,476	2	 <br/>
java.util.WeakHashMap$Entry	28,414	3	727,415	4	 <br/>
java.util.HashMap	27,029	3	691,947	4	 <br/>
java.util.WeakHashMap$Entry[]	19,986	2	194,526	1	 <br/>
java.util.Hashtable$Entry[]	17,852	2	311,216	1	 <br/>
java.util.Hashtable	11,176	1	286,110	1	 <br/>
java.util.WeakHashMap	9,118	1	194,526	1	 <br/>
org.apache.commons.jelly.JellyContext	8,378	1	214,498	1	storage/sw/hudson/war/WEB-INF/lib/commons-jelly-1.1-jenkins-20110627.jar<br/>
org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1	8,090	1	115,058	0	storage/sw/hudson/war/WEB-INF/lib/stapler-jelly-1.187.jar<br/>
java.lang.String[]	7,711	1	174,016	1	 <br/>
java.util.ArrayList	7,655	1	326,640	1</p>


<p>We recently switched from using Jetty as a container for Jenkins to Winstone. But again, I don't know if that is relevant.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=akarollil">akarollil</a>:</i>
<p>This seems to happen only when using Chrome, possibly having auto-refresh enabled and leaving the tab open all the time. </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hjhafner">hjhafner</a>:</i>
<p>We have also Problems with Firefox on Linux and Windows 7 and Internet Explorer on Windows 7.<br/>
After some time IE shows following error:<br/>
<span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/21805/21805_OoM.jpg" style="border: 0px solid black" /></span><br/>
In this case, autorefresh was disabled.<br/>
We observed this memory leak using Jenkins 1.457 and 1.459. Currently we are using 1.454. This version seems to have no problem with memory leaks in browser.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>We also have client-side memory problems. Running Jenkins 1.471, seen with both IE and Firefox. It builds up major memory "waste" overnight, and I'm not using auto-refresh.</p>

<p>I took a snapshot of Firefox's about:memory the other day, after one night of Jenkins (here only the Jenkins compartment is shown):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>1,165.05 MB (100.0%) ++  -- explicit
├────730.22 MB (62.68%) ── heap-unclassified
├────419.95 MB (36.05%) ++  -- js
│    ├──325.76 MB (27.96%) ++  -- compartment(http://jenkins/)
│    │  ├──256.25 MB (22.00%) ++  -- gc-heap
│    │  │  ├──167.00 MB (14.33%) ++  -- objects
│    │  │  │  ├──126.47 MB (10.86%) ── non-function
│    │  │  │  └───40.52 MB (03.48%) ── function
│    │  │  ├───76.72 MB (06.58%) ++  -- shapes
│    │  │  │   ├──64.94 MB (05.57%) ── dict
│    │  │  │   └──11.78 MB (01.01%) ++  -- (2 tiny)
│    │  │  │      ├──11.44 MB (00.98%) ── base
│    │  │  │      └───0.34 MB (00.03%) ── tree
│    │  │  ├───11.92 MB (01.02%) ++  -- arena
│    │  │  │   ├───7.62 MB (00.65%) ── unused
│    │  │  │   ├───2.30 MB (00.20%) ── padding
│    │  │  │   └───2.00 MB (00.17%) ── headers
│    │  │  └────0.61 MB (00.05%) ++  -- (3 tiny)
│    │  │       ├──0.46 MB (00.04%) ── scripts
│    │  │       ├──0.12 MB (00.01%) ── type-objects
│    │  │       └──0.03 MB (00.00%) ── strings
│    │  ├───37.10 MB (03.18%) ++  -- shapes-extra
│    │  │   ├──36.47 MB (03.13%) ── dict-tables
│    │  │   └───0.63 MB (00.05%) ++  -- (3 tiny)
│    │  │       ├──0.50 MB (00.04%) ── tree-tables
│    │  │       ├──0.08 MB (00.01%) ── tree-shape-kids
│    │  │       └──0.05 MB (00.00%) ── compartment-tables
│    │  ├───27.99 MB (02.40%) ── object-slots
│    │  └────4.41 MB (00.38%) ++  -- (7 tiny)
│    │       ├──2.01 MB (00.17%) ── object-elements
│    │       ├──0.97 MB (00.08%) ── analysis-temporary
│    │       ├──0.66 MB (00.06%) ── script-data
│    │       ├──0.45 MB (00.04%) ── mjit-code
│    │       ├──0.25 MB (00.02%) ++  -- type-inference
│    │       │  ├──0.16 MB (00.01%) ── object-main
│    │       │  ├──0.05 MB (00.00%) ── script-main
│    │       │  └──0.03 MB (00.00%) ── tables
│    │       ├──0.05 MB (00.00%) ── mjit-data
│    │       └──0.02 MB (00.00%) ── string-chars
</pre>
</div></div>

<p>Then I did a refresh (F5), which caused the machine+firefox to thrash for 10-15 minutes while it cleaned up its memory. After the dust settled, about:memory said this (again cut to Jenkins):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>146.06 MB (100.0%) ++  -- explicit
├───96.13 MB (65.82%) ++  -- js
│   ├──59.40 MB (40.67%) ── gc-heap-decommitted
│   ├──10.38 MB (07.11%) ++  -- compartment(http://jenkins/)
│   │  ├───6.78 MB (04.64%) ++  -- gc-heap
│   │  │   ├──4.00 MB (02.74%) ++  -- arena
│   │  │   │  ├──3.88 MB (02.66%) ── unused
│   │  │   │  └──0.12 MB (00.08%) ++  -- (2 tiny)
│   │  │   │     ├──0.06 MB (00.04%) ── padding
│   │  │   │     └──0.05 MB (00.04%) ── headers
│   │  │   └──2.78 MB (01.90%) ++  -- (5 tiny)
│   │  │      ├──1.24 MB (00.85%) ++  -- objects
│   │  │      │  ├──0.72 MB (00.50%) ── non-function
│   │  │      │  └──0.51 MB (00.35%) ── function
│   │  │      ├──0.94 MB (00.64%) ++  -- shapes
│   │  │      │  ├──0.45 MB (00.31%) ── dict
│   │  │      │  ├──0.34 MB (00.23%) ── tree
│   │  │      │  └──0.15 MB (00.11%) ── base
│   │  │      ├──0.46 MB (00.31%) ── scripts
│   │  │      ├──0.12 MB (00.08%) ── type-objects
│   │  │      └──0.03 MB (00.02%) ── strings
│   │  └───3.60 MB (02.47%) ++  -- (9 tiny)
│   │      ├──1.10 MB (00.75%) ── analysis-temporary
│   │      ├──0.85 MB (00.58%) ++  -- shapes-extra
│   │      │  ├──0.50 MB (00.34%) ── tree-tables
│   │      │  ├──0.23 MB (00.16%) ── dict-tables
│   │      │  ├──0.08 MB (00.05%) ── tree-shape-kids
│   │      │  └──0.05 MB (00.03%) ── compartment-tables
│   │      ├──0.66 MB (00.45%) ── script-data
│   │      ├──0.45 MB (00.31%) ── mjit-code
│   │      ├──0.26 MB (00.18%) ++  -- type-inference
│   │      │  ├──0.16 MB (00.11%) ── object-main
│   │      │  ├──0.07 MB (00.05%) ── script-main
│   │      │  └──0.03 MB (00.02%) ── tables
│   │      ├──0.20 MB (00.14%) ── object-slots
│   │      ├──0.05 MB (00.04%) ── mjit-data
│   │      ├──0.02 MB (00.01%) ── string-chars
│   │      └──0.01 MB (00.01%) ── object-elements
</pre>
</div></div>

<p>Much better, and I remember to close my browser after work ...</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>Still very much a problem with Jenkins 1.476 ...</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=alex01ves">alex01ves</a>:</i>
<p>1.480.3 - same issue on Chrome.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=frodeaa">frodeaa</a>:</i>
<p>Suspect the memory leak is caused by the <tt>refreshPart</tt> method in hudson-behavior.js <br/>
Heap snapshot shows detached DOM elements that related to the content that are refreshed by the method.</p>

<p>Changing the <tt>refreshPart</tt> to just update <tt>innerHTML</tt> (<tt>hist.innerHTML = rsp.responseText;</tt>) seems to fix "some" of the memory leak.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeHeader panelHeader" style="border-bottom-width: 1px;"><b><a href="https://github.com/jenkinsci/jenkins/blob/master/war/src/main/webapp/scripts/hudson-behavior.js" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/master/war/src/main/webapp/scripts/hudson-behavior.js</a></b></div><div class="codeContent panelContent">
<pre class="code-javascript"><span class="code-comment">// refresh a part of the HTML specified by the given ID,
</span><span class="code-comment">// by using the contents fetched from the given URL.
</span><span class="code-keyword">function</span> refreshPart(id,url) {
    <span class="code-keyword">var</span> f = <span class="code-keyword">function</span>() {
<span class="code-keyword">if</span>(isPageVisible()) {
    <span class="code-keyword">new</span> Ajax.Request(url, {
onSuccess: <span class="code-keyword">function</span>(rsp) {
    <span class="code-keyword">var</span> hist = $(id);
    <span class="code-keyword">if</span> (hist==<span class="code-keyword">null</span>) console.log(<span class="code-quote">"There's no element that has ID of "</span>+id)
    <span class="code-keyword">var</span> p = hist.up();
					
    hist.innerHTML = rsp.responseText;

    <span class="code-comment">//<span class="code-keyword">var</span> div = document.createElement(<span class="code-quote">'div'</span>);
</span>    <span class="code-comment">//div.innerHTML = rsp.responseText;
</span>    <span class="code-comment">//
</span>    <span class="code-comment">//<span class="code-keyword">var</span> node = $(div).firstDescendant();
</span>    <span class="code-comment">//p.replaceChild(node, hist);
</span>    <span class="code-comment">//
</span>    <span class="code-comment">//Behaviour.applySubtree(node);
</span>    <span class="code-comment">//layoutUpdateCallback.call();
</span>
    <span class="code-keyword">if</span>(isRunAsTest) <span class="code-keyword">return</span>;
    refreshPart(id,url);

}
    });    
} <span class="code-keyword">else</span> {
    <span class="code-comment">// Reschedule
</span>    <span class="code-keyword">if</span>(isRunAsTest) <span class="code-keyword">return</span>;
    refreshPart(id,url);
}

    };
    <span class="code-comment">// <span class="code-keyword">if</span> run as test, just <span class="code-keyword">do</span> it once and <span class="code-keyword">do</span> it now to make sure it's working,
</span>    <span class="code-comment">// but don't repeat.
</span>    <span class="code-keyword">if</span>(isRunAsTest) f();
    <span class="code-keyword">else</span>    window.setTimeout(f, 5000);
}</pre>
</div></div>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: lpancescu<br/>
Path:<br/>
 war/src/main/webapp/scripts/hudson-behavior.js<br/>
<a href="http://jenkins-ci.org/commit/jenkins/2f0632ae189a11b439e10a59918434f3248a5787" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/2f0632ae189a11b439e10a59918434f3248a5787</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-10912" title="Browser/JS memory leak in dashboard" class="issue-link" data-issue-key="JENKINS-10912"><del>JENKINS-10912</del></a> Use setInterval in refreshPart (#2539)</p>

<ul>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-10912" title="Browser/JS memory leak in dashboard" class="issue-link" data-issue-key="JENKINS-10912"><del>JENKINS-10912</del></a> Use setInterval in refreshPart</li>
</ul>


<p>The current implementation of refreshPart creates a new closure every 5<br/>
seconds, which, in combination with XMLHttpRequest objects, results in a<br/>
significant memory leak in all major browsers. By using<br/>
window.setInterval to schedule periodic refreshes, only one closure per<br/>
id is created. Please see issue #10912 in the Jenkins tracker for<br/>
further details.</p>

<ul>
	<li>Stop periodical calls if we can't find the div</li>
</ul>


<ul>
	<li>Don't check if isRunAsTest changed after page load</li>
</ul>


<p>(cherry picked from commit 096614a6d72c19633909248cb5b06a179b8b4040)</p>
